Core/Transmog: Add TransmogSets

Golrag opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

-  Add Achievement for transmog set completion
-  SPELL_EFFECT_LEARN_TRANSMOG_SET

**Target branch(es):**
- master

**Issues addressed:** Closes #20133


**Tests performed:**
- Tested in game
- Compiles


**Known issues and TODO list:** (add/remove lines as needed)

<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

Golrag committed:
Core/Transmog: Add TransmogSets
- Achievement
- SPELL_EFFECT_LEARN_TRANSMOG_SET

Golrag committed:
rename left overs.........

Shoxxo commented:
Tested ...

[Working](https://image.prntscr.com/image/33FDOv_KSZSeDM17qVPiRA.png)

Golrag commented:
Oh I am talking about the tab 'Sets' in your appearances collection not about the transmogging to a set.
I could take a look at that tho and see if I it fits this PR

Shoxxo commented:
Yes i know and as you can see, it is avaible or i am wrong?

Golrag commented:
Yea it's available :)

Golrag committed:
Add missing drop table if exists

Shauren commented on src/server/game/Entities/Player/CollectionMgr.cpp:
```diff
@@ -603,6 +603,58 @@ void CollectionMgr::AddItemAppearance(uint32 itemId, uint32 appearanceModId /*=
+    }
+}
+
+bool CollectionMgr::IsSetCompleted(uint32 transmogSetId) const
```
https://gist.github.com/Shauren/6929981f075b8a40b6c822a209fd518a
Improved version with no temporary containers (this is what the client does too btw with ItemTransmogrificationSlots)

Shauren commented on src/server/game/DataStores/DB2Stores.cpp:
```diff
@@ -1958,6 +1978,24 @@ bool DB2Manager::IsToyItem(uint32 toy) const
     return _toys.count(toy) > 0;
 }
 
+std::vector<TransmogSetEntry const*> const* DB2Manager::GetTransmogSetsForItemModifiedAppearance(uint32 itemModID) const
```
Please use full name for argument (itemModifiedAppearanceId)

Golrag committed:
Rename some fields

Move ItemTransmogrificationSlots from item.cpp to item.h
Improve CollectionMgr::IsSetCompleted (thanks Shauren)

Golrag commented on src/server/game/Entities/Player/CollectionMgr.cpp:
```diff
@@ -603,6 +603,50 @@ void CollectionMgr::AddItemAppearance(uint32 itemId, uint32 appearanceModId /*=
+            continue;
+
+        ItemEntry const* item = sItemStore.LookupEntry(itemModifiedAppearance->ItemID);
+        int32 transmogSlot = ItemTransmogrificationSlots[item->InventoryType];
```
@Shauren shouldn't we check if item is null or is it guaranteed to exist ?

Shauren replied:
You should check it - i didn't do complete checking there

Golrag commented:
I could take a look at CRITERIA_TYPE_APPEARANCE_UNLOCKED_BY_SLOT in this PR btw maybe it fits :p

Shauren commented on src/server/game/Entities/Item/Item.h:
```diff
@@ -124,6 +124,39 @@ struct ItemDynamicFieldGems
 };
 #pragma pack(pop)
 
+int32 const ItemTransmogrificationSlots[MAX_INVTYPE] =
```
This change is bad, all you had to do here was write a declaration - `extern int32 const ItemTransmogrificationSlots[MAX_INVTYPE]`

Golrag committed:
Check for Item

also revert changes in item.cpp

Golrag committed:
Fix rebase mistake

Golrag committed:
Add CRITERIA_TYPE_APPEARANCE_UNLOCKED_BY_SLOT

Golrag commented:
Added CRITERIA_TYPE_APPEARANCE_UNLOCKED_BY_SLOT but somehow progress is not shown just like with issue #18448
